addons/udes_advanced_packaging/models/stock_picking.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+from collections import defaultdict
+
+from odoo import models, _
+from odoo.exceptions import ValidationError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _should_reserve_full_packages(self):
+        """Method to determine if picking should reserve entire packages"""
+        self.ensure_one()
+        return self.picking_type_id.u_reserve_as_packages
+
+    def _reserve_full_packages(self):
+        """
+        If the picking type of the picking in self has full package
+        reservation enabled, whole packages will be reserved.
+        """
+        Quant = self.env["stock.quant"]
+        MoveLine = self.env["stock.move.line"]
+
+        # do not reserve full packages when bypass_reserve_full packages
+        # is set in the context as True
+        if not self.env.context.get("bypass_reserve_full_packages"):
+            for picking in self.filtered(lambda p: p._should_reserve_full_packages()):
+                all_quants = Quant.browse()
+                remaining_qtys = defaultdict(int)
+
+                # get all packages
+                packages = picking.mapped("move_line_ids.package_id")
+                for package in packages:
+                    move_lines = picking.mapped("move_line_ids").filtered(
+                        lambda ml: ml.package_id == package
+                    )
+                    pack_products = frozenset(package._get_all_products_quantities().items())
```

Comment:
I think this may cause a mismatch, this line gets the quantity info from the product uom, the below is in terms of the UoM of the move line (potentially different)

Reply:
This was ported from UDES11, is this a suggestion of improvement from existing behaviour, as it seems to be working fine for UDES11

Reply:
It works in 11 as uom doesn't exist. Everything is in eaches (units of 1's).
With different units of measure between the moves and products you may be comparing the quantities of the moves with the quants. 
Example:
if apples come in packs of 2 - and you have 10. You have a quantity of 5 (5*2=10, we never see the 10 in the fields as they all respect the UoM)
If you have a move for a pack of 5 apples- in a UoM of, the product_uom_qty = 1, the product_qty=5. Full pack reservation would be 3 packs of apples - to cover the 5. 

So this isn't in 11 - but this code assumes everywhere the same thing UoM matches. 
You can create an issue to investigate if things need to be changed in this code, change it now, or if you are happy it should work add a unit test to show it :).
If ti is the former, can you add a comment to which bits this impacts and link to the issue. (Even better write a unit test - if it passes awesome no issue, if not add a skip and link to the issue as well :).

Reply:
As discussed elsehwere, let's add TODOs referencing issue 962.

Reply:
Resolving here, both _get_all_products_quantities() should return the same measure, so we add TODOs there.